src/shogun/regression/LeastAngleRegression.cpp

```diff
@@ -172,80 +176,86 @@ bool CLeastAngleRegression::train_machine(CFeatures* data)
-		copy(Xy.begin(), Xy.end(), corr.begin());
-		cblas_dgemv(CblasColMajor, CblasTrans, n_vec, n_fea, -1,
-			X.matrix, n_vec, &mu[0], 1, 1, &corr[0], 1);
-
+		Map<VectorXd> map_corr(&corr[0], n_fea);
+		Map<VectorXd> map_mu(&mu[0], n_vec);
+		
+		map_corr = map_Xy - (map_X.transpose()*map_mu);
+		
 		// corr_sign = sign(corr)
 		for (uint32_t i=0; i < corr.size(); ++i)
 			corr_sign[i] = CMath::sign(corr[i]);
 
 		// find max absolute correlation in inactive set
 		find_max_abs(corr, m_is_active, i_max_corr, max_corr);
-
+		
 		if (!lasso_cond)
 		{
 			// update Cholesky factorization matrix
-			R=cholesky_insert(X, R, i_max_corr);
+			if (m_num_active == 0)
+			{ 
+				// R isn't allocated yet
+				// diag_k = X[:,i_max_corr]' * X[:,i_max_corr]
+				R=SGMatrix<float64_t>(1,1);
+				float64_t diag_k = map_X.col(i_max_corr).dot(map_X.col(i_max_corr));
+				R(0,0) = CMath::sqrt(diag_k);
+			}
+			else
+				R=CLeastAngleRegressionImpl::cholesky_insert_helper(X, X_active, R, i_max_corr, m_num_active);
 			activate_variable(i_max_corr);
 		}
 
-		// corr_sign_a = corr_sign[m_active_set]
-		vector<float64_t> corr_sign_a(m_num_active);
+		// Active variables
+		Map<MatrixXd> map_Xa(X_active, n_vec, m_num_active);
+		if (!lasso_cond)
+			map_Xa.col(m_num_active-1)=map_X.col(i_max_corr);
+		
+		SGVector<float64_t> corr_sign_a(m_num_active);
 		for (int32_t i=0; i < m_num_active; ++i)
 			corr_sign_a[i] = corr_sign[m_active_set[i]];
 
-		// GA1 = R\(R'\corr_sign_a)
-		vector<float64_t> GA1(corr_sign_a);
-		cblas_dtrsm(CblasColMajor, CblasLeft, CblasUpper, CblasTrans, CblasNonUnit,
-			m_num_active, 1, 1, R.matrix, m_num_active, &GA1[0], m_num_active);
-		cblas_dtrsm(CblasColMajor, CblasLeft, CblasUpper, CblasNoTrans, CblasNonUnit,
-			m_num_active, 1, 1, R.matrix, m_num_active, &GA1[0], m_num_active);
+		Map<VectorXd> map_corr_sign_a(corr_sign_a.vector, corr_sign_a.size());
+		Map<MatrixXd> map_R(R.matrix, R.num_rows, R.num_cols);
+		VectorXd solve = map_R.transpose().triangularView<Lower>().solve<OnTheLeft>(map_corr_sign_a);
+		VectorXd GA1 = map_R.triangularView<Upper>().solve<OnTheLeft>(solve);
 
 		// AA = 1/sqrt(GA1' * corr_sign_a)
-		float64_t AA = cblas_ddot(m_num_active, &GA1[0], 1, &corr_sign_a[0], 1);
+		float64_t AA = GA1.dot(map_corr_sign_a);
 		AA = 1/CMath::sqrt(AA);
 
-		// wA = AA*GA1
-		vector<float64_t> wA(GA1);
-		for (int32_t i=0; i < m_num_active; ++i)
-			wA[i] *= AA;
+		VectorXd wA = AA*GA1;
 
 		// equiangular direction (unit vector)
 		vector<float64_t> u = make_vector(n_vec, 0);
-		// u = X[:,m_active_set] * wA
-		for (int32_t i=0; i < m_num_active; ++i)
-		{
-			// u += wA[i] * X[:,m_active_set[i]]
-			cblas_daxpy(n_vec, wA[i],
-				X.get_column_vector(m_active_set[i]), 1, &u[0], 1);
-		}
+		Map<VectorXd> map_u(&u[0], n_vec);
+		
+		map_u=map_Xa*wA;
 
-		// step size
 		float64_t gamma = max_corr / AA;
 		if (m_num_active < n_fea)
 		{
+			#pragma omp parallel for shared(gamma)
 			for (int32_t i=0; i < n_fea; ++i)
```

Comment:
Havent checked exactly how much it helps. Its just parallelizing the dot products.